jina/executors/indexers/cache.py

```diff
@@ -22,35 +24,102 @@ def post_init(self):
+        cached_field = doc.id
+        if self.field == CONTENT_HASH_KEY:
+            cached_field = doc.content_hash
         self._size += 1
-        self.write_handler.write(np.int64(d_id).tobytes())
-
-    def query(self, doc_id: UniqueId, *args, **kwargs) -> Optional[bool]:
-        d_id = int(doc_id)
-        return (d_id in self.query_handler) or None
-
-    def get_query_handler(self):
-        with open(self.index_abspath, 'rb') as fp:
-            return set(np.frombuffer(fp.read(), dtype=np.int64))
-
-    @cached_property
-    def null_query_handler(self):
-        """The empty query handler when :meth:`get_query_handler` fails"""
-        return set()
+        self.query_handler.ids.append(doc.id)
+        self.query_handler.cache.append(cached_field)
+
+    @deprecated_alias(doc_id='doc')
+    def query(self, doc, *args, **kwargs) -> Optional[bool]:
+        """
+        Check whether the given doc's cached field exists in the index
+
+        :param doc: the Document you want to query for
+        """
+        # FIXME
+        if self.query_handler is None:
+            self.query_handler = self.get_query_handler()
+        cached_field = doc.id
+        if self.field == CONTENT_HASH_KEY:
+            cached_field = doc.content_hash
+        status = (cached_field in self.query_handler.cache) or None
+        return status
+
+    def update(self, keys: Iterator[int], values: Iterator[any], *args, **kwargs):
+        """
+        :param keys: list of Document.id
+        :param values: list of either `id` or `content_hash` of :class:`Document"""
+        missed = check_keys_exist(keys, self.query_handler.ids)
+        if missed:
+            raise KeyError(f'Keys {missed} were not found in {self.index_abspath}. No operation performed...')
+
+        for key, doc in zip(keys, values):
+            key_idx = self.query_handler.ids.index(key)
+            cached_field = doc.id
+            if self.field == CONTENT_HASH_KEY:
+                cached_field = doc.content_hash
+            self.query_handler.cache[key_idx] = cached_field
+
+    def delete(self, keys: Iterator[int], *args, **kwargs):
+        """
+        :param keys: list of Document.id
+        """
+        missed = check_keys_exist(keys, self.query_handler.ids)
+        if missed:
+            raise KeyError(f'Keys {missed} were not found. No operation performed...')
+
+        for key in keys:
+            key_idx = self.query_handler.ids.index(key)
+            self.query_handler.ids = [id for idx, id in enumerate(self.query_handler.ids) if idx != key_idx]
+            self.query_handler.cache = [cache for idx, cache in enumerate(self.query_handler.cache) if idx != key_idx]
+            self._size -= 1
 
     def get_add_handler(self):
-        return open(self.index_abspath, 'ab')
+        pass
+
+    def get_query_handler(self) -> CacheHandler:
+        return self.CacheHandler(self.index_abspath, self.logger)
 
     def get_create_handler(self):
```

Comment:
how is then the data created?

Reply:
we don't need a separate handler

Reply:
https://github.com/jina-ai/jina/blob/bbc73f514ee6000e3a7cc1509c2db3294241e53d/jina/executors/indexers/cache.py#L76

Reply:
ok, may u add small comment about this?

Reply:
good point, did